File: Android/app/src/main/java/com/hhp227/datemate/data/datasource/StorageRemoteDataSource.kt
package com.hhp227.datemate.data.datasource

import android.net.Uri
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.tasks.await

class StorageRemoteDataSource private constructor(
    private val storage: FirebaseStorage
) {
    suspend fun uploadFile(uri: Uri, path: String): String {
        val storageRef = storage.reference.child(path)

        storageRef.putFile(uri).await()
        return storageRef.downloadUrl.await().toString()
    }

    companion object {
        @Volatile
        private var instance: StorageRemoteDataSource? = null

        fun getInstance(storage: FirebaseStorage): StorageRemoteDataSource =
            instance ?: synchronized(this) {
                instance ?: StorageRemoteDataSource(storage).also { instance = it }
            }
    }
}

File: iOS/DateMate/Data/DataSource/StorageRemoteDataSource.swift
//
//  StorageRemoteDataSource.swift
//  DateMate
//
//  Created by 홍희표 on 2025/11/26.
//

import Combine
import Foundation
import FirebaseStorage

class StorageRemoteDataSource {
    let storage: Storage
    
    func uploadFile(fileUrl: URL, path: String) -> AnyPublisher<String, Error> {
        Future { promise in
            let storageRef = self.storage.reference().child(path)
            let _ = storageRef.putFile(from: fileUrl, metadata: nil) { _, error in
                if let error = error {
                    promise(.failure(error))
                    return
                }
                storageRef.downloadURL { url, error in
                    if let url = url {
                        promise(.success(url.absoluteString))
                    } else if let error = error {
                        promise(.failure(error))
                    } else {
                        promise(.failure(NSError(domain: "StorageError", code: -1)))
                    }
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    private init(_ storage: Storage) {
        self.storage = storage
    }
    
    private static var instance: StorageRemoteDataSource? = nil
    
    static func getInstance(storage: Storage) -> StorageRemoteDataSource {
        if let instance = self.instance {
            return instance
        } else {
            let storageDataSource = StorageRemoteDataSource(storage)
            self.instance = storageDataSource
            return storageDataSource
        }
    }
}
